drivers: uart: Add support for st half-duplex mode

helo9 opened the pull request:
This PR is a WIP proposal to integrate st's half-duplex uart mode into zephyr.
Thereby rx and tx are connected and only tx pin is used.
Code is put together quickly, to gather feedback on that feature and I did not
put any effort into test implementation, because I am not to sure, where to put it.

erwango reviewed (commented):
Proposal looks good to me (w/o performing an actual review)

I'd just like @dcpleung to confirm this could be limited to STM32 driver as was done for stop bit:
https://github.com/zephyrproject-rtos/zephyr/pull/39427 and there is no impact on uart api.

erwango commented:
> I'd just like @dcpleung to confirm this could be limited to STM32 driver as was done for stop bit:
#39427 and there is no impact on uart api.

^^ @dcpleung

dcpleung commented:
> > I'd just like @dcpleung to confirm this could be limited to STM32 driver as was done for stop bit:
> > #39427 and there is no impact on uart api.

Weird that I never got a notification about this... hm...

Anyway, it looks good. But please add a `sample.yaml` and some comments on which boards the sample can be used on.

gmarull commented on drivers/serial/uart_stm32.c:
```diff
@@ -1685,6 +1690,15 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
 #define UART_DMA_CHANNEL(index, dir, DIR, src, dest)
 #endif
 
+#if !(HAS_LPUART_1)
+	#define STM32_UART_SINGLE_WIRE(index) \
+		.single_wire = DT_INST_PROP(index, single_wire),
+#else
+	#define STM32_UART_SINGLE_WIRE(index) \
+		.single_wire = false,
+#endif
```
can be simply reduced to
```
.single_wire = DT_INST_PROP(index, single_wire),
```

helo9 replied:
Can't follow here. I had trouble with some LPUARTs as they use the same driver, but I did not add the `single-line` property into their bindings _st,stm32-lpuart.yaml_. So compilation fails for each lpuart. To my understanding they do not support this single-line feature, so I didn't want to add it there.

erwango replied:
In that case, can you try the following ?
```
     .single_wire = DT_INST_PROP_OR(index, single_wire, false),
```

gmarull replied:
Yep, what @erwango suggest is what I actually meant sorry :-)

helo9 replied:
Much nicer solution and CI does not complain!

gmarull commented on samples/drivers/uart/single_wire/README.rst:
```diff
@@ -0,0 +1,30 @@
+.. _single_wire:
```
This sample seems to be board-specific, so it should be moved to `samples/boards`. Moreover, please add sample into a separate commit.

helo9 replied:
I would rather try to get hands on some different boards and add further overlays.

helo9 replied:
Separated the commits.

erwango replied:
I agree sample location should be adapted to be more specific.
This being said, having platform specific drivers samples/tests is a common thing (cf `tests/drivers/can`, `tests/drivers/clock_control`,), so I think `samples/drivers/uart/stm32/single_wire` should work.
@gmarull, what do you think about this proposal?

gmarull replied:
As far as it's clear that the sample is specific for STM32 targets I'm fine with it.

helo9 replied:
sounds good, will move the example

helo9 replied:
Done

gmarull commented on samples/drivers/uart/single_wire/app.overlay:
```diff
@@ -0,0 +1,13 @@
+/*
```
use a board overlay, not app.overlay so the sample can be re-used for other boards in the future.

helo9 replied:
done

gmarull commented on samples/drivers/uart/single_wire/src/main.c:
```diff
@@ -0,0 +1,44 @@
+#include <devicetree.h>
+#include <drivers/uart.h>
+
+#define USART2_NODE DT_NODELABEL(usart2)
+#define UART4_NODE DT_NODELABEL(uart4)
+
+const struct device *usart2 = DEVICE_DT_GET(USART2_NODE);
+const struct device *uart4 = DEVICE_DT_GET(UART4_NODE);
```
this makes sample board specific, can be improved (e.g. using aliases?)

helo9 replied:
done

gmarull commented on samples/drivers/uart/single_wire/src/main.c:
```diff
@@ -0,0 +1,44 @@
+
+		uart_poll_out(usart2, 'c');
+
+		k_busy_wait(1000);
```
question: why not k_sleep?

helo9 replied:
no idea anymore, will replace..

helo9 committed:
drivers: uart: Add support for st single wire mode

An additional devicetree poperty `single-wire` is added
to uart and usart bindings of stm32. The driver checks this value
during initialization and enables the single wire mode when set.

Signed-off-by: Jonathan Hahn <Jonathan.Hahn@t-online.de>

helo9 committed:
samples: uart: Add sample for single line uart

The sample uses st single-line uart mode on stm32 boards.

Signed-off-by: Jonathan Hahn <Jonathan.Hahn@t-online.de>

erwango commented on samples/drivers/uart/single_wire/boards/stm32f3_disco.overlay:
```diff
@@ -0,0 +1,20 @@
+
+/ {
+     aliases {
+             sinlge-line-uart1 = &usart2;
```
s/sinlge/single

helo9 replied:
oh

helo9 replied:
fixed

erwango commented on samples/drivers/uart/single_wire/boards/stm32f3_disco.overlay:
```diff
@@ -0,0 +1,20 @@
+/ {
+     aliases {
+             sinlge-line-uart1 = &usart2;
+			 sinlge-line-uart2 = &uart4;
```
Indentation

helo9 replied:
fixed

erwango commented on dts/bindings/serial/st,stm32-uart.yaml:
```diff
@@ -11,6 +11,12 @@ properties:
+      type: boolean
+      required: false
+      description: |
+        Enable the single wire functionality
```
Please provide more details here.
I suggest (just a proposal, don't hesitate to fix based on you experience of the feature)
```
Enables single-wire half-duplex communication. 
Using this mode, TX and RX lines are internally connected, so only TX or RX pin should be configured.
RX/TX conflicts should be handled on user side.
```

helo9 replied:
Sounds good. Never seen the option to choose using TX or RX pin, F1, F3 and H7 always use external TX pin, so adapted in that direction.

erwango commented on samples/drivers/uart/single_wire/boards/stm32f3_disco.overlay:
```diff
@@ -0,0 +1,20 @@
+	};
+};
+
+&usart2 {
+	single-wire;
+};
```
Can't we also update pinctrl-0 here and configure only one pin ?

helo9 replied:
Never changed anything there, but I'll look into that.

helo9 replied:
Updated binding description makes this mandatory.

helo9 replied:
Done

erwango commented on samples/drivers/uart/single_wire/sample.yaml:
```diff
@@ -0,0 +1,6 @@
+sample:
+  name: STM32 Single Wire UART sample
+tests:
+  sample.drivers.uart.single_wire:
```
Update sample reference when moving it.

helo9 replied:
Thanks for the reminder ;)

erwango commented on samples/drivers/uart/single_wire/sample.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  sample.drivers.uart.single_wire:
+    platform_allow: stm32f3_disco
+    tags: drivers uart
```
Would be nice to add a harness to check the sample is behaving correctly in CI.
Cf hello world sample.

helo9 replied:
yeah, I was also thinking about that, but did not figure out, how to do that. Already asked on discord https://discord.com/channels/720317445772017664/733037944922964069/925822535533223957, but did not get an answer so far. Will have a look at hello world sample

PerMac replied:
Do I understand correctly from the question at discord, that this test requires an STM board with a physical connection between certain pins? Upstream CI is only able of executing tests/samples on emulated/simulated platforms and only build for physical boards. As for harnesses: Most samples use console harness (as hello_word) where the console output is verified according to what is in a yaml. There is also ztest harness allowing to use asserts in a code. 
If the physical connection between the pins is required you need to add "fixture" field. E.g https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/uart/uart_async_api/testcase.yaml

helo9 replied:
I have added the fixture and described its usage in the README.

erwango commented on samples/drivers/uart/stm32/single_wire/boards/stm32f3_disco.overlay:
```diff
@@ -0,0 +1,20 @@
+};
+
+&usart2 {
+	single-wire;
```
Not functionally  required, but educating.
```suggestion
	pinctrl-0 = <&usart2_tx_pa2>;
	single-wire;
```

helo9 replied:
ah didn't look into it yet, but now I get what you meant.. I thought of something in c code which disallows usage of anything more than the tx pin

helo9 replied:
done

erwango commented on samples/drivers/uart/stm32/single_wire/boards/stm32f3_disco.overlay:
```diff
@@ -0,0 +1,20 @@
+	single-wire;
+};
+
+&uart4 {
```
```suggestion
&uart4 {
	pinctrl-0 = <&uart4_tx_pc10>;
```

helo9 replied:
done

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1701,6 +1706,7 @@ static const struct uart_stm32_config uart_stm32_cfg_##index = {	\
 	.hw_flow_control = DT_INST_PROP(index, hw_flow_control),	\
 	.parity = DT_INST_ENUM_IDX_OR(index, parity, UART_CFG_PARITY_NONE),	\
 	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),			\
+	.single_wire = DT_INST_PROP_OR(index, single_wire, false), \
```
Similarly than `hw_flow_control`, booleans are `false` if not present.
So just use `DT_INST_PROP` instead of `DT_INST_PROP_OR`

helo9 replied:
the problem here is, that this property might not be defined within the binding, so I guess the OR is needed, see https://github.com/zephyrproject-rtos/zephyr/pull/40525#discussion_r777373128

erwango commented on drivers/serial/uart_stm32.h:
```diff
@@ -24,6 +24,7 @@ struct uart_stm32_config {
 	/* initial parity, 0 for none, 1 for odd, 2 for even */
 	int  parity;
 	const struct pinctrl_dev_config *pcfg;
+	bool single_wire;
```
Please move this one rank above, together with `hw_flow_control` and `parity`.
Also add a comment (you can mention half-duplex for instance)

helo9 replied:
Done

helo9 replied:
Done

erwango commented on dts/bindings/serial/st,stm32-uart.yaml:
```diff
@@ -18,7 +18,7 @@ properties:
         Using this mode, TX and RX lines are internally connected and
         only TX pin is used afterwards and should be configured.
-        RX/TX conflicts should be handled on user side.
+        RX/TX conflicts must be handled on user side.
```
Please move this to previous commit.

helo9 replied:
oh, yeah messed something up there. Done.

erwango commented on dts/bindings/serial/st,stm32-usart.yaml:
```diff
@@ -18,7 +18,7 @@ properties:
         Using this mode, TX and RX lines are internally connected and
         only TX pin is used afterwards and should be configured.
-        RX/TX conflicts should be handled on user side.
+        RX/TX conflicts must be handled on user side.
```
Ditto

helo9 replied:
Done

erwango commented on samples/drivers/uart/stm32/single_wire/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(uart_basic_api)
```
```suggestion
project(uart_stm32_single_wire)
```

helo9 replied:
Done

erwango commented on samples/drivers/uart/stm32/single_wire/README.rst:
```diff
@@ -0,0 +1,33 @@
+.. _single_wire:
```
s/single_wire/stm32_single_wire

helo9 replied:
Done

erwango commented on samples/drivers/uart/stm32/single_wire/README.rst:
```diff
@@ -0,0 +1,33 @@
+.. _single_wire:
+
+Single Wire UART
```
This will show up directly here: https://docs.zephyrproject.org/latest/samples/boards/boards.html.
Please rename as `STM32 Single Wire UART`

helo9 replied:
Done

erwango commented on samples/drivers/uart/stm32/single_wire/src/main.c:
```diff
@@ -0,0 +1,45 @@
+		int ret = uart_poll_in(sl_uart2, &recv);
+
+		if (ret < 0) {
+			printk("receive on uart4 failed with %d\n", ret);
```
Let's replace `uart4` with something neutral for sample genericity, or get node label printed here.

helo9 replied:
Oh sure. Done.

erwango reviewed (changes requested):
Thanks @helo9, last round of comments.

erwango commented on samples/drivers/uart/stm32/single_wire/README.rst:
```diff
@@ -0,0 +1,33 @@
+.. s/single_wire/stm32_single_wire:
```
To be clear, just do:
```suggestion
.. _sample_uart_stm32_single_wire:
```

helo9 replied:
Sure, I am still wondering anyways, where this stuff is used. Assumed it was something sphinx specific, but did not see something fitting, when scanning the sphinx rest documentation.

helo9 replied:
Adapted.

helo9 commented on samples/drivers/uart/stm32/single_wire/src/main.c:
```diff
@@ -0,0 +1,45 @@
+		int ret = uart_poll_in(sl_uart2, &recv);
+
+		if (ret < 0) {
+			printk("Receiving failed with %d\n", ret);
```
That's better, adapted.

erwango commented:
@gmarull Would you mind having a new look ?
